generator/views/funnel_analysis_view.py

```diff
@@ -0,0 +1,238 @@
+"""
+from __future__ import annotations
+
+from typing import Any, Dict, Iterator, List, Optional
+
+from .view import View, ViewDict
+
+DEFAULT_NUM_FUNNEL_STEPS: int = 4
+
+
+class FunnelAnalysisView(View):
+    """A view for doing Funnel Analysis."""
+
+    type: str = "funnel_analysis_view"
+
+    def __init__(self, namespace: str, tables: List[Dict[str, str]]):
+        """Get an instance of a FunnelAnalysisView."""
+        super().__init__(namespace, "funnel_analysis", FunnelAnalysisView.type, tables)
+
+    @classmethod
+    def from_db_views(
+        klass,
+        namespace: str,
+        is_glean: bool,
+        channels: List[Dict[str, str]],
+        db_views: dict,
+        num_funnel_steps: int = DEFAULT_NUM_FUNNEL_STEPS,
+    ) -> Iterator[FunnelAnalysisView]:
+        """Get Client Count Views from db views and app variants.
+
+        We only create a FunnelAnalysisView if we have the two necessary db tables:
+            - events_daily
+            - event_types
+        """
+        # We can guarantee there will always be at least one channel,
+        # because this comes from the associated _get_glean_repos in
+        # namespaces.py
+        dataset = next(
+            (channel for channel in channels if channel.get("channel") == "release"),
+            channels[0],
+        )["dataset"]
+
+        necessary_views = {"events_daily", "event_types"}
+        actual_views = {}
+        for view_id, references in db_views[dataset].items():
+            if view_id in necessary_views:
+                actual_views[view_id] = f"`mozdata.{dataset}.{view_id}`"
+
+        if len(actual_views) == 2:
+            # Only create an instance if we have the two necessary tables
+            tables = {
+                "funnel_analysis": "events_daily_table",
+                "event_types": actual_views["event_types"],
+            }
+            tables.update(
+                {
+                    f"event_type_{i}": "event_types"
+                    for i in range(1, num_funnel_steps + 1)
+                }
+            )
+            yield FunnelAnalysisView(
+                namespace,
+                [tables],
+            )
+
+    @classmethod
+    def from_dict(
+        klass, namespace: str, name: str, _dict: ViewDict
+    ) -> FunnelAnalysisView:
+        """Get a FunnalAnalysisView from a dict representation."""
+        return FunnelAnalysisView(namespace, _dict["tables"])
+
+    def to_lookml(self, bq_client, v1_name: Optional[str]) -> Dict[str, Any]:
+        """Get this view as LookML."""
+        return {
+            "includes": [f"{self.tables[0]['funnel_analysis']}.view.lkml"],
+            "views": self._funnel_analysis_lookml() + self._event_types_lookml(),
+        }
+
+    def n_events(self) -> int:
+        """Get the number of events allowed in this funnel."""
+        return len([k for k in self.tables[0] if k.startswith("event_type_")])
+
+    def _funnel_analysis_lookml(self) -> List[Dict[str, Any]]:
```

Comment:
This seems like the core of this explore. Taking it on faith here that this is debugged and works well :)

Reply:
Yup, this is based on the [UJET funnel analysis explore](https://github.com/mozilla/looker-spoke-default/blob/main/user_journey/views/funnel_analysis/funnel_analysis.view.lkml). That has been sufficiently debugged and used.